### PR TITLE
Add more testings and modify the source

### PR DESCRIPTION
**Pull request by peter279k**
Firstly, thank you for your OpenLoad PHP client!
The changed log is as follows:
# Changed log
- add PHP 7.1 testing in ```travis.yml```.
- check the array key ```headers``` whether it is existed.
- If the file is not existed, it should throw the ```InvalidArgumentException```.
- set the file sha1 hash value.
- the ```__toString``` should return raw JSON string.
- setFileId and ```__toString``` (```TickBuilder``` class)
- remove the setter/getter setFileId/getFileId method in ```Ticket``` class(it's useless)

I'm not sure that approaches are appropriate.
Please review this and suggest the recommendation and we can discuss this.

Thanks.

### CODE REVIEW

**Comment by Ideneal** [comment]
Hi Peter thank you for support my project.

I agree with some points but I disagree with others.
I try to explain my disagrees about some points as well as possible in the following list:

- The class constructor with `$data` parameter must be removed. I believe that is useless, it's not a good practice and it is not my style.
- The implementation of all `__toString` methods must be reset to the previous state. Those you've removed are usefull and those you've added it's not a good practice and not a good pattern (`json_encode` should be used on external classes that have a formatter behaviour)
- The methods `setFileId`/`getFileId` is usefull for anyone would want to know at which file correspond a ticket. So reset it to the previous state.
- The array `headers` check must be removed because the `headers` is an optional parameter.

Can you fix this issues?

**Comment by peter279k** [comment]
Hi @Ideneal, I agree with your suggestions.
If possible, can you add the comment for that code you want to modify?
And I can review them conveniently.

Thanks.

**Comment by Ideneal** [comment]
Ok perfect :+1:

File: src/Builder/ContentBuilder.php [hunk 1]
```diff
@@ -49,10 +49,11 @@ public static function buildFile(array $data)
     {
         $uploadDate = (new \DateTime())->setTimestamp($data['upload_at']);
 
-        $file = new File();
+        $file = new File($data);
```
Comment (Ideneal) [inline src/Builder/ContentBuilder.php]: Remove `$data` argument

File: src/Builder/FileInfoBuilder.php [hunk 2]
```diff
@@ -29,7 +29,7 @@ class FileInfoBuilder extends AbstractBuilder
      */
     public static function buildFileInfo(array $data)
     {
-        $fileInfo = new FileInfo();
+        $fileInfo = new FileInfo($data);
```
Comment (Ideneal) [inline src/Builder/FileInfoBuilder.php]: Remove `$data` argument

File: src/Builder/LinkBuilder.php [hunk 3]
```diff
@@ -30,7 +30,7 @@ class LinkBuilder extends AbstractBuilder
      */
     public static function buildDownloadLink(array $data)
     {
-        $downloadLink = new DownloadLink();
+        $downloadLink = new DownloadLink($data);
```
Comment (Ideneal) [inline src/Builder/LinkBuilder.php]: Remove `$data` argument.

File: src/Builder/LinkBuilder.php [hunk 4]
```diff
@@ -52,7 +52,7 @@ public static function buildDownloadLink(array $data)
      */
     public static function buildUploadLink(array $data)
     {
-        $uploadLink = new UploadLink();
+        $uploadLink = new UploadLink($data);
```
Comment (Ideneal) [inline src/Builder/LinkBuilder.php]: Remove `$data` argument.

File: src/Builder/TicketBuilder.php [hunk 5]
```diff
@@ -29,7 +29,7 @@ class TicketBuilder extends AbstractBuilder
      */
     public static function buildTicket(array $data)
     {
-        $ticket  = new Ticket();
+        $ticket  = new Ticket($data);
```
Comment (Ideneal) [inline src/Builder/TicketBuilder.php]: Remove `$data` argument.

File: src/Entity/AbstractContent.php [hunk 6]
```diff
@@ -73,14 +73,4 @@ public function setName($name)
         $this->name = $name;
         return $this;
     }
-
-    /**
-     * To string method
-     *
-     * @return string
-     */
-    public function __toString()
```
Comment (Ideneal) [inline src/Entity/AbstractContent.php]: This should remain.

File: src/Entity/AbstractLink.php [hunk 7]
```diff
@@ -45,14 +45,4 @@ public function setUrl($url)
         $this->url = $url;
         return $this;
     }
-
-    /**
-     * To string method
-     *
-     * @return string
-     */
-    public function __toString()
```
Comment (Ideneal) [inline src/Entity/AbstractLink.php]: This should remain.

File: src/Entity/DownloadLink.php [hunk 8]
```diff
@@ -49,6 +49,16 @@ class DownloadLink extends AbstractLink
     private $token;
 
     /**
+     * @var array The download link information
```
Comment (Ideneal) [inline src/Entity/DownloadLink.php]: Remove this.

File: src/Entity/DownloadLink.php [hunk 9]
```diff
@@ -185,4 +195,14 @@ public function setToken($token)
         $this->token = $token;
         return $this;
     }
+
```
Comment (Ideneal) [inline src/Entity/DownloadLink.php]: Remove this

File: src/Entity/File.php [hunk 10]
```diff
@@ -64,6 +64,16 @@ class File extends AbstractContent
     private $url;
 
     /**
+     * @var string The file information
```
Comment (Ideneal) [inline src/Entity/File.php]: Remove this

File: src/Entity/File.php [hunk 11]
```diff
@@ -269,4 +279,14 @@ public function setUrl($url)
         $this->url = $url;
         return $this;
     }
+
```
Comment (Ideneal) [inline src/Entity/File.php]: Remove this

File: src/Entity/FileInfo.php [hunk 12]
```diff
@@ -49,6 +49,16 @@ class FileInfo
     private $contentType;
 
     /**
+     * @var array The file information data
```
Comment (Ideneal) [inline src/Entity/FileInfo.php]: Remove this

File: src/Entity/FileInfo.php [hunk 13]
```diff
@@ -193,6 +203,6 @@ public function setContentType($contentType)
      */
     public function __toString()
     {
-        return $this->id;
```
Comment (Ideneal) [inline src/Entity/FileInfo.php]: Maintain the previous code than newest.

File: src/Entity/Ticket.php [hunk 14]
```diff
@@ -39,9 +39,14 @@ class Ticket
     private $expirationDate;
 
     /**
-     * @var string The file id related to the ticket
```
Comment (Ideneal) [inline src/Entity/Ticket.php]: Maintain the old code than newest

File: src/Entity/Ticket.php [hunk 15]
```diff
@@ -84,16 +89,6 @@ public function getExpirationDate()
     }
 
     /**
-     * Returns the file id related to the ticket
```
Comment (Ideneal) [inline src/Entity/Ticket.php]: Maintain the old code

File: src/Entity/Ticket.php [hunk 16]
```diff
@@ -146,25 +141,12 @@ public function setExpirationDate(\DateTime $expirationDate)
     }
 
     /**
-     * Sets the file id related to the ticket
```
Comment (Ideneal) [inline src/Entity/Ticket.php]: Maintain the old code

File: src/Entity/Ticket.php [hunk 17]
```diff
@@ -146,25 +141,12 @@ public function setExpirationDate(\DateTime $expirationDate)
...
-    }
-
-    /**
      * To string method
      *
      * @return string
      */
     public function __toString()
     {
-        return $this->code;
```
Comment (Ideneal) [inline src/Entity/Ticket.php]: Maintain the old code

File: src/Entity/UploadLink.php [hunk 18]
```diff
@@ -45,4 +55,14 @@ public function setExpirationDate(\DateTime $expirationDate)
         $this->expirationDate = $expirationDate;
         return $this;
     }
+
```
Comment (Ideneal) [inline src/Entity/UploadLink.php]: Remove this

File: src/OpenLoadClient.php [hunk 19]
```diff
@@ -225,6 +225,8 @@ public function uploadRemoteFile($url, $folder = null, array $headers = array())
             $params['folder'] = (string) $folder;
         }
 
+        array_key_exists('headers', $params) ? : $params['headers'] = '';
```
Comment (Ideneal) [inline src/OpenLoadClient.php]: Remove this

File: src/OpenLoadClient.php [hunk 20]
```diff
@@ -516,6 +518,10 @@ public function getVideoSplashImage($file)
      */
     public function uploadFile($fileName, $folder = null, $httpOnly = false)
     {
+        if (!file_exists($fileName)) {
+            throw new \InvalidArgumentException('The'.$fileName.'is not found!');
```
Comment (Ideneal) [inline src/OpenLoadClient.php]: Change the message into: 'The file '. $fileName .' does not exist.'

File: tests/Builder/ContentBuilderTest.php [hunk 21]
```diff
@@ -48,5 +48,17 @@ public function testBuildFile()
...
+        $this->assertEquals('4258', $file->getFolderId());
+        $this->assertInstanceOf('\DateTime', $file->getUploadDate());
+        $this->assertEquals('active', $file->getStatus());
+        $this->assertEquals('5114011', $file->getSize());
+        $this->assertEquals('video/mp4', $file->getContentType());
+        $this->assertEquals('48', $file->getDownloadCount());
+        $this->assertEquals('ok', $file->getConversionStatus());
+        $this->assertEquals('UPPjeAk--30', $file->getId());
+        $this->assertEquals('big_buck_bunny.mp4.mp4', $file->getName());
+        $this->assertEquals(json_encode($data), $file);
```
Comment (Ideneal) [inline tests/Builder/ContentBuilderTest.php]: Remove the assertEquals of object with json.

File: tests/Builder/FileInfoBuilderTest.php [hunk 22]
```diff
@@ -33,5 +33,12 @@ public function testBuildFileInfo()
         $data     = json_decode($this->fixture, true);
         $fileInfo = FileInfoBuilder::buildFileInfo($data);
         $this->assertInstanceOf('Ideneal\OpenLoad\Entity\FileInfo', $fileInfo);
+        $this->assertEquals('72fA-_Lq8Ak3', $fileInfo->getId());
+        $this->assertEquals('200', $fileInfo->getStatus());
+        $this->assertEquals('The quick brown fox.txt', $fileInfo->getName());
+        $this->assertEquals(123456789012, $fileInfo->getSize());
+        $this->assertEquals('2fd4e1c67a2d28fced849ee1bb76e7391b93eb12', $fileInfo->getSha1());
+        $this->assertEquals('plain/text', $fileInfo->getContentType());
+        $this->assertEquals(json_encode($data), $fileInfo);
```
Comment (Ideneal) [inline tests/Builder/FileInfoBuilderTest.php]: Remove the assertEquals of object with json.

File: tests/Builder/LinkBuilderTest.php [hunk 23]
```diff
@@ -38,6 +38,9 @@ public function testBuildUploadLink()
         $data       = json_decode($this->uploadFixture, true);
         $uploadLink = LinkBuilder::buildUploadLink($data);
         $this->assertInstanceOf('Ideneal\OpenLoad\Entity\UploadLink', $uploadLink);
+        $this->assertInstanceOf('\DateTime', $uploadLink->getExpirationDate(new \DateTime('2011-01-26 13:33:37')));
+        $this->assertEquals(json_encode($data), $uploadLink);
```
Comment (Ideneal) [inline tests/Builder/LinkBuilderTest.php]: Remove the assertEquals of object with json.

File: tests/Builder/LinkBuilderTest.php [hunk 24]
```diff
@@ -48,5 +51,13 @@ public function testBuildDownloadLink()
         $data         = json_decode($this->downloadFixture, true);
         $downloadLink = LinkBuilder::buildDownloadLink($data);
         $this->assertInstanceOf('Ideneal\OpenLoad\Entity\DownloadLink', $downloadLink);
+        $this->assertEquals('The quick brown fox.txt', $downloadLink->getName());
+        $this->assertEquals(12345, $downloadLink->getSize());
+        $this->assertEquals('2fd4e1c67a2d28fced849ee1bb76e7391b93eb12', $downloadLink->getSha1());
+        $this->assertEquals('plain/text', $downloadLink->getContentType());
+        $this->assertInstanceOf('\DateTime', $downloadLink->getUploadDate());
+        $this->assertEquals('4spxX_-cSO4', $downloadLink->getToken());
+        $this->assertEquals(json_encode($data), $downloadLink);
```
Comment (Ideneal) [inline tests/Builder/LinkBuilderTest.php]: Remove the assertEquals of object with json.

File: tests/Builder/TicketBuilderTest.php [hunk 25]
```diff
@@ -34,6 +34,11 @@ public function testBuildTicket()
         $data   = json_decode($this->fixture, true);
         $ticket = TicketBuilder::buildTicket($data);
         $this->assertInstanceOf('Ideneal\OpenLoad\Entity\Ticket', $ticket);
+        $this->assertEquals('72fA-_Lq8Ak~~1440353112~n~~0~nXtN3RI-nsEa28Iq', $ticket->getCode());
+        $this->assertInstanceOf('Ideneal\OpenLoad\Entity\Captcha', $ticket->getCaptcha());
+        $this->assertEquals(10, $ticket->getWaitTime());
+        $this->assertInstanceOf('\DateTime', $ticket->getExpirationDate());
+        $this->assertEquals(json_encode($data), $ticket->__toString());
```
Comment (Ideneal) [inline tests/Builder/TicketBuilderTest.php]: Remove the assertEquals of object with json.

**Comment by peter279k** [comment]
Hi @Ideneal, thank you for your requests. I've modified the code that you want to change completely.
Please review this.
One more thing you mentioned that I update the dependencies in ```composer.json``` and add other testings in the existed PHP file.

Thanks.

File: src/Entity/Ticket.php [hunk 26]
```diff
@@ -39,11 +39,32 @@ class Ticket
     private $expirationDate;
 
     /**
-     * @var string The file id related to the ticket
+     * @var array string The file id related to the ticket
```
Comment (Ideneal) [inline src/Entity/Ticket.php]: It is not an array but is a string

File: src/OpenLoadClient.php [hunk 27]
```diff
@@ -516,6 +516,10 @@ public function getVideoSplashImage($file)
      */
     public function uploadFile($fileName, $folder = null, $httpOnly = false)
     {
+        if (!file_exists($fileName)) {
+            throw new \InvalidArgumentException('The'.$fileName.'does not exist.');
```
Comment (Ideneal) [inline src/OpenLoadClient.php]: Add spaces after 'The' and before 'does'

**Pull request merged**